google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java

```diff
@@ -3273,4 +3273,45 @@ public void testBlobReload() throws Exception {
     updated.delete();
     assertNull(updated.reload());
   }
+
+  private Blob createBlob(String method, BlobInfo blobInfo) throws IOException {
+    switch (method) {
+      case "create":
+        return storage.create(blobInfo);
+      case "writer":
+        {
+          storage.writer(blobInfo).close();
+          return storage.get(BlobId.of(blobInfo.getBucket(), blobInfo.getName()));
+        }
+      default:
+        throw new IllegalArgumentException("Unknown method " + method);
+    }
+  }
+
+  private void testAutoContentType(String method) throws IOException {
+    String[] names = {"file1.txt", "dir with spaces/Pic.Jpg", "no_extension"};
```

Comment:
It's not always possible. In case of resumeable upload a blob is created first and the content is available later:

```
writer = storage.writer(blobInfo); // creating a blob, detecting content-type, no contect
writer.write(content);
writer.close();              
```

I did an experiment:
I created a file `PDF.txt` with a pdf content. Then I uploaded it as `PDF.jpg` with various clients and checked the resulting content type. I got the following:

gsutil:  text/plain
NodeJS:  image/jpeg
Go: application/pdf

To achieve consistent behavior across all the clients it should be implemented on the server-side at the moment when the upload is finished.

Until this feature is implemented I suggest type detecting based on file extension, as it has been done recently in nodejs (https://github.com/googleapis/nodejs-storage/pull/1190)